cookbooks/aws-parallelcluster-config/files/default/custom_action_executor/custom_action_executor.py

```diff
@@ -0,0 +1,518 @@
+        # preserving error case for making the script executable
+        try:
+            subprocess.run(["chmod", "+x", exe_script.path], check=True)  # nosec - trusted input
+        except subprocess.CalledProcessError as err:
+            raise DownloadRunError(
+                f"Failed to run {self.event_name} script {exe_script.step_num} {exe_script.url} "
+                f"due to a failure in making the file executable, return code: {err.returncode}.",
+                f"Failed to run {self.event_name} script {exe_script.step_num} "
+                f"due to a failure in making the file executable, return code: {err.returncode}.",
+            ) from err
+
+        # execute script with it's args
+        try:
+            # arguments are provided by the user who has the privilege to create/update the cluster
+            subprocess.run([exe_script.path] + (exe_script.args or []), check=True)  # nosec - trusted input
+        except subprocess.CalledProcessError as err:
+            raise DownloadRunError(
+                f"Failed to execute {self.event_name} script {exe_script.step_num} {exe_script.url},"
+                f" return code: {err.returncode}.",
+                f"Failed to execute {self.event_name} script {exe_script.step_num}, return code: {err.returncode}.",
+            ) from err
+
+    @staticmethod
+    def _is_https_url(url):
+        return urlparse(url).scheme == "https"
+
+    @staticmethod
+    def _is_s3_url(url):
+        return urlparse(url).scheme == "s3"
+
+    @staticmethod
+    def _parse_s3_url(url):
+        parsed_url = urlparse(url)
+        return parsed_url.netloc, parsed_url.path.lstrip("/")
+
+
+class LegacyEventName(Enum):
+    """Maps legacy events names to avoid changing script contract."""
+
+    ON_NODE_START = "preinstall"
+    ON_NODE_CONFIGURED = "postinstall"
+    ON_NODE_UPDATED = "postupdate"
+
+    def map_to_current_name(self):
+        """Return the current event name value as it's configured in the cluster config."""
+        if self == LegacyEventName.ON_NODE_START:
+            result = "OnNodeStart"
+        elif self == LegacyEventName.ON_NODE_CONFIGURED:
+            result = "OnNodeConfigured"
+        elif self == LegacyEventName.ON_NODE_UPDATED:
+            result = "OnNodeUpdated"
+        else:
+            raise ValueError(f"Unknown legacy event name: {self.value}")
+
+        return result
+
+    def __str__(self):
+        """Return the legacy event name value."""
+        return self.value
+
+
+@dataclass
+class CustomActionsConfig:
+    """
+    Encapsulates custom actions configuration.
+
+    Contains all the configuration relevant to custom actions execution.
+    """
+
+    stack_name: str
+    region_name: str
+    node_type: str
+    queue_name: str
+    event_name: str
+    legacy_event: LegacyEventName
+    can_execute: bool
+    dry_run: bool
+    script_sequence: list  # type list[ScriptDefinition]
```

Comment:
The situation for list types in Python is:
- Prior to 3.9 you can import capital case `List` from the `typing` module and specify the type of objects in the list with `List[ScriptDefinition]`.
- From 3.9 onward you can use `list[ScriptDefinition]` and the `List` form is deprecated, as the lower-case version supports subscripting (see [doc](https://docs.python.org/3/library/typing.html#typing.List))